src/dmd/iasmgcc.d

```diff
@@ -0,0 +1,362 @@
+ *      |                                : OutputOperands
+ *      |                                [ : InputOperands
+ *      |                                [ : Clobbers ] ]; }
+ *      | GotoAsmStatement:
+ *      |     asm [FunctionAttributes] { Expression
+ *      |                                :
+ *      |                                : InputOperands
+ *      |                                : Clobbers
+ *      |                                : GotoLabels; }
+ * Params:
+ *      p = parser state
+ *      s = asm statement to parse
+ * Returns:
+ *      the parsed gcc asm statement
+ */
+
+GccAsmStatement parseGccAsm(Parser)(Parser p, GccAsmStatement s)
+{
+    s.insn = p.parseExpression();
+    if (p.token.value == TOK.semicolon)
+        goto Ldone;
+
+    // No semicolon followed after instruction template, treat as extended asm.
+    for (int section = 0; section < 4; ++section)
+    {
+        p.check(TOK.colon);
+
+        final switch (section)
+        {
+            case 0:
+                s.outputargs = p.parseExtAsmOperands(s);
+                break;
+
+            case 1:
+                p.parseExtAsmOperands(s);
+                break;
+
+            case 2:
+                s.clobbers = p.parseExtAsmClobbers();
+                break;
+
+            case 3:
+                s.labels = p.parseExtAsmGotoLabels();
+                break;
+        }
+
+        if (p.token.value == TOK.semicolon)
+            goto Ldone;
+    }
+Ldone:
+    p.check(TOK.semicolon);
+
+    return s;
+}
+
+/***********************************
+ * Parse and run semantic analysis on a GccAsmStatement.
+ * Params:
+ *      s  = gcc asm statement being parsed
+ *      sc = the scope where the asm statement is located
+ * Returns:
+ *      the completed gcc asm statement, or null if errors occurred
+ */
+public Statement extendedAsmSemantic(GccAsmStatement s, Scope *sc)
+{
+    //printf("GccAsmStatement.semantic()\n");
+    scope p = new Parser!ASTCodegen(sc._module, ";", false);
+
+    // Make a safe copy of the token list before parsing.
+    Token *toklist = null;
+    Token **ptoklist = &toklist;
+
+    for (Token *token = s.tokens; token; token = token.next)
+    {
+        *ptoklist = Token.alloc();
+        memcpy(*ptoklist, token, Token.sizeof);
+        ptoklist = &(*ptoklist).next;
+        *ptoklist = null;
+    }
+    p.token = *toklist;
+
+    // Parse the gcc asm statement.
+    s = p.parseGccAsm(s);
+    if (p.errors)
+        return null;
+    s.stc = sc.stc;
+
+    // Fold the instruction template string.
+    if (s.insn.op == TOK.mixin_)
+        s.insn = (cast(CompileExp)s.insn).e1;
+    s.insn = semanticString(sc, s.insn, "asm instruction template");
+
+    if (s.labels && s.outputargs)
+        s.error("extended asm statements with labels cannot have output constraints");
+
+    // Analyse all input and output operands.
+    if (s.args)
+    {
+        for (size_t i = 0; i < s.args.dim; i++)
```

Comment:
Could be a foreach

Reply:
If you mean `foreach (i; 0 .. s.args.dim)`, then ok.